src/validate.rs

```diff
@@ -1,33 +1,1490 @@
  */
-pub fn validate(
+pub fn validate_translation(
     config: LanguageConfig,
     base: String,
     case: String,
     translation: String,
-) -> Option<ValidationError> {
-    None
+) -> ValidationResult {
+    let base = match ParsedString::parse(&base) {
+        Err(_) => {
+            return ValidationResult {
+                errors: vec![ValidationError {
+                    critical: true,
+                    position: None,
+                    message: String::from("Base language text is invalid."),
+                    suggestion: Some(String::from("This is a bug; wait until it is fixed.")),
+                }],
+                normalized: None,
+            };
+        }
+        Ok(parsed) => parsed,
+    };
+    if case != "" && case != "default" {
+        if !config.allow_cases() {
+            return ValidationResult {
+                errors: vec![ValidationError {
+                    critical: true,
+                    position: None,
+                    message: String::from("No cases allowed."),
+                    suggestion: None,
+                }],
+                normalized: None,
+            };
+        } else if !config.cases.contains(&case) {
+            return ValidationResult {
+                errors: vec![ValidationError {
+                    critical: true,
+                    position: None,
+                    message: format!("Unknown case '{}'.", case),
+                    suggestion: Some(format!("Known cases are: '{}'", config.cases.join("', '"))),
+                }],
+                normalized: None,
+            };
+        }
+    }
+    let mut translation = match ParsedString::parse(&translation) {
+        Err(err) => {
+            return ValidationResult {
+                errors: vec![ValidationError {
+                    critical: true,
+                    position: Some(err.position),
+                    message: err.message,
+                    suggestion: None,
+                }],
+                normalized: None,
+            };
+        }
+        Ok(parsed) => parsed,
+    };
+    let errs = validate_string(&config, &translation, Some(&base));
+    if errs.iter().any(|e| e.critical) {
+        ValidationResult {
+            errors: errs,
+            normalized: None,
+        }
+    } else {
+        sanitize_whitespace(&mut translation);
+        normalize_string(&config.get_dialect(), &mut translation);
+        ValidationResult {
+            errors: errs,
+            normalized: Some(translation.compile()),
+        }
+    }
+}
+
+fn remove_ascii_ctrl(t: &mut String) {
+    *t = t.replace(|c| char::is_ascii_control(&c), " ");
+}
+
+fn remove_trailing_blanks(t: &mut String) {
+    if let Some(last) = t.rfind(|c| c != ' ') {
+        t.truncate(last + 1);
+    }
+}
+
+/// Replace all ASCII control codes with blank.
+/// Remove trailing blanks at end of each line.
```

Comment:
Is this meant as doxygen-style comment? Or a reminder to yourself?

Otherwise, can I suggest `/** .. */` instead?

Reply:
This is meant as `rustdoc`, which seems to use `///`.